framework/src/tinyalsa/tinyalsa.c

```diff
@@ -531,49 +535,99 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
-#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
-#else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
-	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
-#endif
+
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
+	if ((!pcm->draining) && (!pcm->running) && (pcm_start(pcm) < 0)) {
 		return -errno;
-	} else {
-		pcm->running = 1;
 	}
 
-	/* Wait for deque message from kernel */
-	size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
-	if (size != sizeof(msg)) {
-		/* Interrupted by a signal? What to do? */
-		return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
-	}
-	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
-		apb = (struct ap_buffer_s *)msg.u.pPtr;
-		/* Copy data to user buffer */
-		memcpy(data, apb->samp, apb->nbytes);
-		/* Enque buffer for next read opertion */
-		nbytes = apb->nbytes;
-		apb->nbytes = 0;
-		apb->curbyte = 0;
-		apb->flags = 0;
-		bufdesc.u.pBuffer = apb;
-		if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
-			return oops(pcm, errno, "failed to enque buffer after read");
+	/* Read data */
+	if (pcm->next_buf != NULL) {
+		/* We have a pending buffer with some data in it. */
+		apb = pcm->next_buf;
+
+		if (nbytes < pcm->next_size) {
+			/* User wants less data than we have in buffer.
+			   Copy the requested amount of data to user buffer and return */
+			memcpy(data, apb->samp + pcm->next_offset, nbytes);
+			pcm->next_size -= nbytes;
+			pcm->next_offset += nbytes;
+		} else {
+			/* User wants more data than we have in buffer.
+			   Copy available data to user buffer, enqueue buffer and return */
+			memcpy(data, apb->samp + pcm->next_offset, pcm->next_size);
+
+			nbytes = pcm->next_size;
+			pcm->next_size = 0;
+			pcm->next_offset = 0;
+			pcm->next_buf = NULL;
+
+			apb->nbytes = 0;
+			apb->curbyte = 0;
+			apb->flags = 0;
+			bufdesc.u.pBuffer = apb;
+			if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
+				return oops(pcm, errno, "failed to enque buffer after read");
+			}
 		}
+
 	} else {
-		return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
+		/* We got here because we dont have any pending buffers */
+		/* Wait for deque message from kernel */
+		if (pcm->draining) {
```

Comment:
pcm_draining will be called after writei  finished,
you don't have to check it in readi..
Also, draining is not neccessary if it always called after writei.